modules/go/main.go

```diff
@@ -2,28 +2,104 @@ package main
 
 import (
 	"context"
+	"fmt"
 	"path"
+	"strings"
 
 	"go.opentelemetry.io/otel/codes"
 	"golang.org/x/sync/errgroup"
 
+	"github.com/containerd/platforms"
 	"github.com/dagger/dagger/modules/go/internal/dagger"
 )
 
+const (
+	defaultPlatform = dagger.Platform("")
+)
+
 func New(
 	// Project source directory
 	source *dagger.Directory,
 	// Go version
 	// +optional
 	// +default="1.23.2"
 	version string,
-) *Go {
+	// Use a custom module cache
+	// +optional
+	moduleCache *dagger.CacheVolume,
+
+	// Use a custom build cache
+	// +optional
+	buildCache *dagger.CacheVolume,
+
+	// Use a custom base container.
+	// The container must have Go installed.
+	// +optional
+	base *dagger.Container,
+
+	// Pass arguments to 'go build -ldflags''
+	// +optional
+	ldflags []string,
+
+	// Add string value definition of the form importpath.name=value
+	// Example: "github.com/my/module.Foo=bar"
+	// +optional
+	values []string,
+
+	// Enable CGO
+	// +optional
+	cgo bool,
+
+	// Enable race detector. Implies cgo=true
+	// +optional
+	race bool,
+) Go {
 	if source == nil {
 		source = dag.Directory()
 	}
-	return &Go{
-		Version: version,
-		Source:  source,
+	if moduleCache == nil {
+		// Cache volumes should be namespaced by module, but they aren't (yet).
+		// For now, we namespace them explicitly here.
+		moduleCache = dag.CacheVolume("github.com/dagger/dagger/modules/go:modules")
+	}
+	if buildCache == nil {
+		// Cache volumes should be namespaced by module, but they aren't (yet).
+		// For now, we namespace them explicitly here.
+		buildCache = dag.CacheVolume("github.com/dagger/dagger/modules/go:build")
+	}
+	if base == nil {
```

Comment:
I wonder if we should still allow this.

We need the base to be customizable (we had a discussion somewhere before about Ubuntu being built on Ubuntu, alpine on alpine, wolfi on wolfi - otherwise we hit weird libc issues). But now we have the additional constraint that this has to have `protoc` installed.

I'm not convinced about the overlay concept here as well in terms of our need to (eventually) do multi platform images here - how do we ensure it works with Ubuntu/alpine/wolfi?

Maybe we could use a distro enum, like we do at the top-level module?

Not a blocker, but would like to discuss the overlay concept more separately in that pr.

Reply:
The `overlay` feature is definitely up for discussion (not a coincidence that it's in a separate PR :)

But `base` I think is a must-have. It is very easy to understand, and opens a very useful escape hatch without much additional complexity. If you bring your own, you're responsible for making sure go is installed. Kind of a no-brainer IMO.

Reply:
Sounds good.

Could we add a sanity check that `base` has the right things installed before we start to use it? It could check the `go version` is the right one, make sure that `protoc` is installed, etc.

Fine as a follow-up, just on a quest to remove footguns!